File: Generic Sorting in C & Haskell/generic_sorting.c
/* CECS 342 – Assignment 5
   Name: Krrish Kohli
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct{
    char name[32];
    int age;
} Person;

//Using this function for qsort.
static int double_asc(const void *a, const void *b)
{
    double da = *(const double*)a; 
    double db = *(const double*)b;
    return (da > db) - (da < db);
}

// sort persons by alphabetical order of their names.
static int person_name_asc(const void *a, const void *b)
{
    const Person *pa = (const Person*)a;
    const Person *pb = (const Person*)b;
    return strcmp(pa->name, pb->name);
}

// sort persons by descending order of their ages if tie sort by ascending order of their name.
static int person_age_desc_name_asc(const void *a, const void *b)
{
    //(const Person*)a/b is cast the untyped pointer a to a pointer to const Person
    const Person *pa = (const Person*)a;
    const Person *pb = (const Person*)b;

    if(pa->age != pb->age)
        return (pb->age - pa->age);
    return strcmp(pa->name, pb->name);
}

//Print array of the numbers
static void print_numbers(const double *xs, size_t n) {
    for (size_t i = 0; i < n; ++i) {
        printf("%s%.2f", (i ? ", " : ""), xs[i]);
    }
    printf("\n");
}


//Print array of people
static void print_people(const Person *ps, size_t n) {
    for (size_t i = 0; i < n; ++i) {
        printf("%s%s, %d", (i ? "; " : ""), ps[i].name, ps[i].age);
    }
    printf("\n");
}

File: Generic Sorting in C & Haskell/GenericSorting.hs
-- GenericSorting.hs
--
-- CECS 342 Assignment 5
-- Krrish Kohli

module GenericSorting where

import Data.List (sort, sortBy, sortOn, intercalate)
import Data.Ord (comparing, Down(..))
import Text.Printf (printf)

--Person type
data Person = Person {name::String, age::Int} deriving (Eq, Show)

--Data Numbers
numbers :: [Double]
numbers = [ 645.41,    37.59,   76.41, 
              5.31, (-34.23),    1.11, 
              1.10,    23.46,  635.47, 
         (-876.32),   467.83,   62.25]

--Data People
people :: [Person]
people =   [ Person "Hal" 20, Person "Susann" 31, Person "Dwight" 19, Person "Kassandra" 21
  , Person "Lawrence" 25, Person "Cindy" 22, Person "Cory" 27, Person "Mac" 19
  , Person "Romana" 27, Person "Doretha" 32, Person "Danna" 20, Person "Zara" 23
  , Person "Rosalyn" 26, Person "Risa" 24, Person "Benny" 28, Person "Juan" 33
  , Person "Natalie" 25]


--Task numbers ascending
sortNumbersAscending :: [Double] -> [Double]
sortNumbersAscending = sort

--Task name ascending
sortPeopleByName :: [Person] -> [Person]
sortPeopleByName = sortOn name

--Task age descending and name ascending
sortPeopleByAgeDescendingNameAscending :: [Person] -> [Person]
sortPeopleByAgeDescendingNameAscending = sortBy (comparing (Down . age) <> comparing name)

-- intercalate ", " ["1.00","2.00","3.00"] == "1.00, 2.00, 3.00"
pNumbers :: [Double] -> String
pNumbers xs = intercalate ", " (map (\x -> printf "%.2f" x) xs) -- map each Double to a String, then "join" with ", "

-- Print a Person as "Name, Age"
pPerson :: Person -> String
pPerson (Person n a) = n ++ ", " ++ show a

-- intercalate "; " ["Hal, 20","Susann, 31"] == "Hal, 20; Susann, 31"
pPeople :: [Person] -> String
pPeople ps = intercalate "; " (map pPerson ps)